### DISCUSSION TITLE
style-loader as part of an npm package?

### DISCUSSION THREAD

**Discussion by danieldelcore** [webpack/webpack#16482] [Q&A] [answered]
Hi there 👋, possibly a bit of a rookie question. 

I'm wondering if it's possible for an NPM package to depend on `style-loader` without the consuming application also having to configure it? 

I'm hoping to use `lazyStyleTag` as part of a package in order to lazy-load CSS files at runtime. 
My assumption is that webpack is responsible for building the package and ensuring that the assets that are imported are part of the `dist`, however, does it also inject some sort of runtime into the package to give it the ability to inject the CSS files? 

I'm currently on `webpack@4.41.3` and `style-loader@1.0.0`. 

Here's an example: 

```jsx
// my-package
const myStyles = require('./my-css-file.css');

export const mountCSS = () => {
  myStyles.use();
}
```

Then in my app: 

```jsx
import { mountCSS } from 'my-package';

mountCSS(); // 🎉
```

I receive the following error: 

```
VM317:1 Uncaught ReferenceError: require is not defined
    at eval (eval at ./my-package/dist/esm/mount-css.js
    at options.factory (react refresh:6:1)
    at __webpack_require__ (bootstrap:24:1)
    at fn (hot module replacement:62:1)
    at ./src/App.js (bundle.js:20:74)
    at options.factory (react refresh:6:1)
    at __webpack_require__ (bootstrap:24:1)
    at fn (hot module replacement:62:1)
    at ./src/index.js (App.js:219:1)
```

Have I completely misinterpreted how webpack should work, or have I just done something wrong in my setup?

**Comment by alexander-akait** [comment]
Why you need this? Can you describe?

**Reply by danieldelcore** [reply]
Hey @alexander-akait, we have multiple "theme" css files that we want to import and mount to the DOM when a user calls our library. 
However, not all themes need to be on the page, only the ones composed by the user, relevant to that particular app. 

The hope was to lazyload these files in at run-time, to avoid loading a bunch of irrelevant styles on the page. The logic to do this exists in an NPM package that consumers would download and use.

**Comment by danieldelcore** [comment, marked as answer]
A bit of an update, I found that there was a problem with the relative URL I was using. I've fixed that now, and it's functional without the lazyloading config.

But when I try to use `use()` it complains about it not existing:

<img width="576" alt="image" src="https://user-images.githubusercontent.com/3030010/202588173-3f758ecc-ee17-4a88-9d27-136161c32c84.png">

**Reply by alexander-akait** [reply]
Can you create small reproducible test repo? I will look and it will be faster to help you

**Reply by danieldelcore** [reply]
Sure thing @alexander-akait, here's an example: https://github.com/danieldelcore/style-loader-repro/tree/main

What I'm looking for is if this repo was published to NPM, should i still be able to share this functionality without Webpack being used by the consuming app?

**Reply by alexander-akait** [reply]
So do you want to use `style-loader` to build libraries? I think better to use vanilla js code for this, because `style-loader` generate some exatra stuff for webpack, also you will have extra runtime code from webpack

**Reply by danieldelcore** [reply]
Ohk that seems sensible. Are there any libraries that can do this, or even a guide to implementing this myself? I've not been able to find one, Google mostly points me to `style-loader`.

**Reply by alexander-akait** [reply]
Just use browser API :smile: 

Anyway if you want to work with styles I recommend do not include styles inside your components, because it is good practice to chnage stylistics of components, just ask developers to include your styles (you can provide example of code for bundler and non bundlers envs), also you can provide example how to override built-in styles and etc

**Reply by danieldelcore** [reply]
Thanks for your help, that's good advice. We're going to take another approach now. Thanks again 👋
